### ISSUE TITLE
audio_output_device and audio_input_device are mixed up (Opensuse TW, Pipewire/Jack)

### ISSUE THREAD

**Issue by zigmhount** [BespokeSynth/BespokeSynth#230]
Hi! I'm really looking forward to give Bespoke a fair trial, it seems similar to what I've been attempting to build using Pure Data, but much sleeker :)
However, while the first start (pre-compiled binary) using Alsa works fine, I then set the devicetype to JACK so that it can use Pipewire, and I can't manage to get Bespoke to initialize audio devices. It actually looks like it is expecting output devices as inputs and inputs as outputs, see screenshots:
![image](https://user-images.githubusercontent.com/58641590/134312294-689caf5d-6d21-4836-91e1-5c50cb798ce6.png)
Output selection dropdown menu:
![image](https://user-images.githubusercontent.com/58641590/134312337-d2803eca-366f-441a-b480-b0c76f5b7a81.png)
Input selection dropdown menu:
![image](https://user-images.githubusercontent.com/58641590/134312375-12d8e435-6b69-49cd-b34f-872b205f4456.png)

Outputs include "... Digital Microphone" whereas inputs include "... Speaker + Headphones".

Trying to manually set the proper input and inputs in `userprefs.json` returns the following errors:
```
$:~/Software/BespokeSynth> ./BespokeSynth
'o != NULL' failed at ../pipewire-jack/src/pipewire-jack.c:4152 jack_port_name()
'destination_port != NULL' failed at ../pipewire-jack/src/pipewire-jack.c:4621 jack_connect()
'o != NULL' failed at ../pipewire-jack/src/pipewire-jack.c:4152 jack_port_name()
'destination_port != NULL' failed at ../pipewire-jack/src/pipewire-jack.c:4621 jack_connect()
```

So it really looks like a mismatch between inputs and outputs' names.

Let me know if additional information may be useful!

Cheers!

**Comment by KottV** [comment]
Hi,

Can you test it with Jackd instead of PipeWire? I maintain that openSUSE build and have use it with Jack/ALSA only.

**Comment by zigmhount** [comment]
Hi! Oh, which openSUSE build are you talking about? I've been using the binary executable from the [zip file](https://github.com/awwbees/BespokeSynth/releases/tag/v1.0.0), just realizing now that it is available on https://software.opensuse.org/package/BespokeSynth.
I'll try that one later today as well as with Jackd on another TW machine (I don't really want to uninstall Pipewire just now on my work laptop).

**Comment by KottV** [comment]
yes, that one is recent https://build.opensuse.org/package/show/home%3Akill_it%3AJUCE/BespokeSynth
but as today TW is in stage of rebuild with glibc-3.34 , so we need to wait until it synced :(

**Comment by zigmhount** [comment]
Yep just tried now :(

**Comment by zigmhount** [comment]
Hm, so, on one laptop with Opensuse TW using Pipewire, installing BespokeSynth from kill_it:JUCE gives the same problem, crashing as soon as I select inputs and outputs for Jack and working fine with Alsa. On the other laptop, also TW with Pipewire, but also a bunch of packages from geekositalia:daw, I installed the same version of BespokeSynth (not the outdated one from geekositalia) and it does work without problem when selecting Jack inputs and outputs while running Pipewire!

The funny thing is that at the same time, I've been struggling for a few weeks to get Bitwig (installed via flatpak) to run with Jack/Pipewire on the 2nd machine, while it works without problem on the 1st laptop...
I guess I'll have to thoroughly compare the versions and vendors of pipewire-libjack and other things, and/or try removing all packages from geekositalia on the 2nd machine, with a risk of BespokeSynth not working anymore but Bitwig working. I'll keep you posted.

**Comment by zigmhount** [comment]
Well, it actually seems to be an issue with Pipewire on OpenSuse. Bespoke works fine with Pipewire 0.3.37 on Ubuntu Studio (I had to compile to use libpython3.9 instead of 3.8).

On OpenSuse Tumbleweed, same Pipewire version, Bespoke runs on one machine and fails on the other (inverted inputs and outputs) when using the Jack backend with Pipewire.
On Opensuse, the midi inputs and outputs are inverted in Bitwig on both laptops with Tumbleweed, causing Bitwig to not be able to connect to the audio server at all - but it works fine as long as no MIDI controllers are connected.
I deactivated the geekositalia repos and did a `zypper dup --allow-vendor-change --allow-downgrade` with pretty much the same repositories on both OpenSuse laptops, so I'm not sure what more to investigate. I might just give up for now and fall back to Jack :(

**Comment by KottV** [comment]
The mos recent version is in https://software.opensuse.org//download.html?project=home%3Abespokesynth&package=bespokesynth now, GDAW package will be synced with soon.

I suppose that it's a PipeWire thing. Unfortunately i'm not using it, i'll ask people in GeekosDAW channel to check.
BTW how the other JUCE standalone apps works in your setup?

**Comment by zigmhount** [comment]
As for other JUCE apps, I just tested with Odin2 from kill_it repo on Opensuse TW, and it crashes when I select Pipewire instead of Alsa. Interesting pattern here.
Bespoke version 1.0.1 also still crashes when selecting Jack with Pipewire on Opensuse Tumbleweed on one laptop. I migrated the other laptop from Opensuse to Ubuntu Studio, where both Bitwig and Bespoke seem for work fine with Pipewire. I haven't found yet a deb package for Odin2 standalone.
